node/network/bridge/src/lib.rs

```diff
@@ -282,37 +324,211 @@ fn construct_view(live_heads: &[Hash]) -> View {
+		NetworkBridgeEvent::OurViewChange(new_view.clone()),
+		ctx,
+	).await {
+		log::warn!(target: TARGET, "Aborting - Failure to dispatch messages to overseer");
+		return Err(e)
+	}
+
+	if let Err(e) = dispatch_collation_event_to_all(
+		NetworkBridgeEvent::OurViewChange(new_view.clone()),
+		ctx,
+	).await {
+		log::warn!(target: TARGET, "Aborting - Failure to dispatch messages to overseer");
+		return Err(e)
+	}
+
+	Ok(())
+}
+
+// Handle messages on a specific peer-set. The peer is expected to be connected on that
+// peer-set.
+async fn handle_peer_messages<M>(
+	peer: PeerId,
+	peers: &mut HashMap<PeerId, PeerData>,
+	messages: Vec<WireMessage<M>>,
+	net: &mut impl Network,
+) -> SubsystemResult<Vec<NetworkBridgeEvent<M>>> {
+	let peer_data = match peers.get_mut(&peer) {
+		None => {
+			net.report_peer(peer, UNCONNECTED_PEERSET_COST).await?;
+
+			return Ok(Vec::new());
+		},
+		Some(d) => d,
+	};
+
+	let mut outgoing_messages = Vec::with_capacity(messages.len());
+	for message in messages {
+		outgoing_messages.push(match message {
+			WireMessage::ViewUpdate(new_view) => {
+				if new_view.0.len() > MAX_VIEW_HEADS {
+					net.report_peer(
+						peer.clone(),
+						MALFORMED_VIEW_COST,
+					).await?;
+
+					continue
+				} else if new_view == peer_data.view {
+					continue
+				} else {
+					peer_data.view = new_view;
+
+					NetworkBridgeEvent::PeerViewChange(
+						peer.clone(),
+						peer_data.view.clone(),
+					)
+				}
+			}
+			WireMessage::ProtocolMessage(message) => {
+				NetworkBridgeEvent::PeerMessage(peer.clone(), message)
+			}
+		})
+	}
+
+	Ok(outgoing_messages)
+}
+
+async fn send_validation_message<I>(
+	net: &mut impl Network,
+	peers: I,
+	message: WireMessage<protocol_v1::ValidationProtocol>,
+) -> SubsystemResult<()>
+	where I: IntoIterator<Item=PeerId>, I::IntoIter: ExactSizeIterator
+{
+	send_message(net, peers, PeerSet::Validation, message).await
+}
+
+async fn send_collation_message<I>(
+	net: &mut impl Network,
+	peers: I,
+	message: WireMessage<protocol_v1::CollationProtocol>,
+) -> SubsystemResult<()>
+	where I: IntoIterator<Item=PeerId>, I::IntoIter: ExactSizeIterator
+{
+	send_message(net, peers, PeerSet::Collation, message).await
+}
 
-	let notifications = peers.keys().cloned()
-		.map(move |peer| Ok(NetworkAction::WriteNotification(peer, message.clone())));
+async fn send_message<M: Encode + Clone, I>(
+	net: &mut impl Network,
+	peers: I,
+	peer_set: PeerSet,
+	message: WireMessage<M>,
+) -> SubsystemResult<()>
+	where I: IntoIterator<Item=PeerId>, I::IntoIter: ExactSizeIterator,
+{
+	let mut message_producer = stream::iter({
+		let peers = peers.into_iter();
+		let n_peers = peers.len();
+		let mut message = Some(message.encode());
+
+		peers.enumerate().map(move |(i, peer)| {
+			// optimization: avoid cloning the message for the last peer in the
+			// list. The message payload can be quite large. If the underlying
+			// network used `Bytes` this would not be necessary.
```

Comment:
I like the mechanism of this optimization, but I don't understand why the `Bytes` type would make this unnecessary. It's not necessary for the purpose of this PR, but it would be nice if you could link me to the type you mean, so I could better understand this statement.

Reply:
this is not new code from the PR, it's just been moved around

Reply:
Check out https://docs.rs/bytes/0.5.6/bytes : it's a fancy `Arc<Vec<u8>>`